src/manual.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2017 Chris Kowalczyk <ckowalczyk@suse.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "manual.h"
+
+#include "transport.h"
+#include "ticket.h"
+#include "config.h"
+#include "log.h"
+#include "request.h"
+
+
+int manual_selection(struct ticket_config *tk,
+	struct booth_site *preference, int update_term, cmd_reason_t reason)
+{
+	if (local->type != SITE)
+		return 0;
+
+	tk_log_debug("starting manual selection");
+	tk_log_debug("selection caused by %s %s",
+				state_to_string(reason),
+				reason == OR_AGAIN ? state_to_string(tk->election_reason) : "" );
+
+	// Manual selection is done without any delay, the leader is assigned
+	set_leader(tk, local);
+	set_state(tk, ST_LEADER);
+
+	// Manual tickets never expire, we don't specify expiration time
+
+	// Make sure that election_end field is empty
+	time_reset(&tk->election_end);
+
+	if (is_time_set(&tk->delay_commit) && all_sites_replied(tk)) {
+		time_reset(&tk->delay_commit);
+		tk_log_debug("reset delay commit as all sites replied");
+	}
+
+	save_committed_tkt(tk);
+
+	// Inform others about the new leader
+	ticket_broadcast(tk, OP_HEARTBEAT, OP_ACK, RLT_SUCCESS, 0);
+	tk->ticket_updated = 0;
+
+	return 0;
+}
```

Comment:
From the implementation viewpoint, this is an exception to normal ticket management, i.e. the ticket is granted synchronously. Maybe that's fine, but maybe it would be beneficial to go through the usual asynchronous path, i.e. ticket_cron().

Reply:
I was thinking about that. However, in case of manual tickets we don't have distinction like for automatic tickets. For example, leader for automatic ticket would expect acks or send heartbeats. A candidate would wait for the votes.
What could apply for a manual ticket, which is being granted? No matter what was its state before, the ticket could only call 'complete selection' function, which would contain the second part of this function.
We could also do something like that: at the beginning of selection set it to 'candidate', wait a while, and set it to 'leader'. Would you prefer something like this?

Reply:
I really cannot say. We'd need to review what all housekeeping tasks are being done for the normal tickets and then see which of those apply for manual too and finally weigh on what is mor practical. The danger of having two paths is that it could happen that in future both of them may need to be updated. Maybe we just leave it as is for now and review a bit later again, once all other bits are in place. Also, IIRC, Yan was talking about honoring certain delay parameter even for manual grants, but cannot recall which one.

Reply:
For manual tickets we want to still honor __acquire_after__ time, which is not really used for granting tickets. This time only tells Booth when it should check what is happening with the ticket.
I think the the whole election for manual tickets (which I called _manual selection_ in code), do not have to bother with any delays, while granting manual tickets, so we do not have to use all those ticket_cron() mechanism here.
However, if you want, I can add this step, as mentioned in my previous comment. But I rather think we don't need it :)

Reply:
Not sure how much of the existing logic would be possible/worth to be reused. But of course it'd be beneficial to somehow unify the basic workflow on handling all kinds of tickets  and potentially simply the maintenance of the code.

The case of granting a granted ticket is closer to the existing workflow I guess. But of course, it still depends on how we'd like to handle it. For example, would it be more convenient to use the existing workflow to honor "acquire_after"?  Could the case possible/necessary to either be split into two steps, revoke&grant?

Reply:
Well, now IIUC, re-grant is not the combination of revoke & grant, right? If so, it's not necessary to honor acquire_after I assume.

Reply:
I wonder when would re-grant (i.e. grant an already granted ticket) happen. Is there a plausible scenario for that? Given that booth doesn't have a configuration reload ... well, what actually happens on booth start when it sees a manual ticket in the configuration?

Reply:
re-grant shouldn't be very often. I can imagine two scenarios:
1) One site with granted ticket is down. Admin grants this ticket to another site. (it is not possible to revoke the ticket because the site is down).
2) Ticket is granted to a site. And _for some strange reason_ admin grants the same ticket to another site. We are ending up having a ticket granted to two sites.

Regarding reading the configuration.. There is not much difference between manual and automatic tickets during startup. We just set a flag for the ticket, really nothing more.

Reply:
Sorry for the confusion. It was the scenario of "move a granted ticket to somewhere else" in my mind ;-) But it seems it's not able to do a "move" as a batch operation yet.

Reply:
Yan, if I recall correctly, I was asked once to implement the "move" command, but argued against it.

Reply:
Sounds somewhat tricky ;-)

Reply:
On the one hand, it would be helpful to have 'move' command, on the other hand when you think what impact it could have on many underlying resources.. perhaps it is better to not make it easier for the user to move tickets :)

Reply:
I'd say so too.